### PR TITLE
implement array_dims

### PR DESCRIPTION
**Pull request by valkrypton**
Added support for `array_dims` postgres function under [#4153](https://github.com/diesel-rs/diesel/issues/4153)
This is my first PR ever so I would love some poitners!

### CODE REVIEW

**Comment by valkrypton** [comment]
I can't fix the doc test i've written. Help would be appreciated

File: diesel/src/pg/expression/functions.rs [hunk 1]
```diff
@@ -748,3 +748,34 @@ define_sql_function! {
...
+    /// #   use diesel::sql_types::{Nullable,Array,Integer};
+    /// #   let connection = &mut establish_connection();
+    ///
+    /// let dims = diesel::select(array_dims::<Array<_>,_,_>(vec![1,2]))
+    ///     .get_result::<String>(connection).unwrap();
+    /// assert!(String::from("[1:2]").eq(&dims));
+    ///
+    /// let dims = diesel::select(array_dims::<Array<_>,_,_>(vec![vec![1,2,3],vec![4,5,6]]))
+    /// .get_result::<String>(connection).unwrap();
+    /// assert!(String::from("[1:2][1:3]").eq(&dims));
```
Comment (guissalustiano) [inline diesel/src/pg/expression/functions.rs:775]: The crazy error here say that diesel didn't know how to transform `Vec<Vec<_>>` to 'Array<_>` becase Diesel does not support multidimensional arrays yet. So you can remove this test for now.

File: diesel/src/pg/expression/functions.rs [hunk 2]
```diff
@@ -748,3 +748,34 @@ define_sql_function! {
...
+    /// #
+    /// # fn main(){
+    /// #    run_test().unwrap();
+    /// # }
+    /// # fn run_test()->QueryResult<()>{
+    /// #   use diesel::dsl::array_dims;
+    /// #   use diesel::sql_types::{Nullable,Array,Integer};
+    /// #   let connection = &mut establish_connection();
+    ///
+    /// let dims = diesel::select(array_dims::<Array<_>,_,_>(vec![1,2]))
```
Comment (guissalustiano) [inline diesel/src/pg/expression/functions.rs:769]:
```suggestion
    /// let dims = diesel::select(array_dims::<Array<Integer>,_,_>(vec![1,2]))
```
The error says that rust don't know with implementation use, this `i32` could be `Int4`, `Int8` or something else, to fix you need to specify the type here. The `array_append` doesn't need this because it Array inner type is the same of the second argument, and the second argument is specified

**Review by guissalustiano — CHANGES_REQUESTED** [review]
Thanks for open the PR!

I let some comment about the errors. Could you also add a test with a None input to check it behavior? Thanks!

**Review by weiznich — APPROVED** [review]

**Comment by weiznich** [comment]
Thanks for working on this. The PR looks now good :+1:

File: diesel/src/pg/expression/functions.rs (hunk 2 above)
Comment (valkrypton) [reply diesel/src/pg/expression/functions.rs]: Thanks for the help!

**Comment by valkrypton** [comment]
@weiznich I also resolved the conflicts and merged the branches, is that a problem?

**Comment by weiznich** [comment]
@valkrypton I's fine, I just tend to resolve conflicts on accepted merge conflicts on my own to not burden contributors with the merge order.

**Comment by valkrypton** [comment]
@weiznich I would love to contribute more to the repo, how do you recommend I start learning about Diesel internals?

**Comment by weiznich** [comment]
@valkrypton Well the best way to learn about internals is to contribute and ask questions as soon as you hit a problem. It is also usful to join the reviewers teams (see #1186) and read code written by others. That is often an excellent place to learn more about how others approach certain problems and you often get explanations there on how certain things are implemented.
